File: Daily task/2463-H-Minimum-Total-Distance-Traveled/min_total_dist_moved.py
class Solution:
    def minimumTotalDistance(self, robot: List[int], factory: List[List[int]]) -> int:
        factory.sort(key=lambda x:x[0]) # sort the factories positions
        robot.sort()
        factory_slots = []
        for fact in factory:
            temp = [fact[0]] * fact[1]
            factory_slots.extend(temp)
        
        n = len(robot)
        m = len(factory_slots)

        OPT = [[0] * (m+1) for _ in range(n+1)]
        for i in range(1, n+1):
            OPT[i][0] = float('inf')
        
        for i in range(1, n+1):
            for j in range(1, m+1):
                OPT[i][j] = min(OPT[i][j-1], OPT[i-1][j-1] + abs(robot[i-1] - factory_slots[j-1]))
        
        return OPT[n][m]

File: Daily task/2463-H-Minimum-Total-Distance-Traveled/min_total_dist_moved.cpp
class Solution {
public:
    long long minimumTotalDistance(vector<int>& robot, vector<vector<int>>& factory) {
        sort(robot.begin(), robot.end());
        sort(factory.begin(), factory.end());

        int n = robot.size();
        
        vector<int> factory_flat; 
        for (int i = 0; i < factory.size(); i++) {
            for (int j = 0; j < factory[i][1]; j++) {
                factory_flat.push_back(factory[i][0]);
            }
        }

        int m = factory_flat.size();

        long long INF = 1e15; 
        vector<vector<long long>> OPT(n + 1, vector<long long>(m + 1, 0));

        for (int i = 1; i < n + 1; i++) {
            OPT[i][0] = INF;
        }

        for (int i = 1; i < n + 1; i++) {
            for (int j = 1; j < m + 1; j++) {
                OPT[i][j] = min(OPT[i][j-1], OPT[i-1][j-1] + abs(robot[i-1] - factory_flat[j-1]));
            }
        }

        return OPT[n][m];
    }
};